ghiscoding/Angular-Slickgrid
# BREAKING CHANGE: use Slickgrid-Universal monorepo next major 3.0

### Next Major Version `3.x` (WIP)

This will be, for the most part, deleting a lot of code and instead use it from the [Slickgrid-Universal](https://github.com/ghiscoding/slickgrid-universal) lib to have 1 common/core lib. The advantage of that extra lib is that it's built as a monorepo and with that we'll be able to decouple a few Angular-Slickgrid Services (Export to File/Excel, OData, GraphQL, ...). So you will end up with a much smaller bundle since you will be bundling only what you really install and use (without having for example OData/GraphQL that you might never use). 

On the migration side, it should be fairly simple following certain steps, the main changes are in regards to the Services that get decoupled, you can take a sneak peak at the changes in the upcoming [Migration Guide](/ghiscoding/Angular-Slickgrid/wiki/Migration-to-3.x)

ETA 🤞 - Sometime in July or early August 🌞⛱️  

###  [Migration Guide](/ghiscoding/Angular-Slickgrid/wiki/Migration-to-3.x)

#### Minimum Requirement Changes
- Angular-Slickgrid `>= 12.1.0`
- ngx-translate `>= 13.0.0`
- TypeScript `>= 4.3.5`
- **RxJS** >=`7.2.0` (also a `major` bump)
- Bootstrap `>= 4.5.0` (or any other UI framework like Bulma)
- IE11 support will be **dropped**, min target will be ES2018

#### List of TODOs
- [x] 1. replace all Formatters with Slickgrid-Universal
- [x] 2. replace all Grouping Formatters with Slickgrid-Universal
- [x] 3. replace all Aggregators with Slickgrid-Universal
- [x] 4. replace all Editors with Slickgrid-Universal
- [x] 5. replace all Editor Valitors with Slickgrid-Universal
- [x] 6. replace all Filters with Slickgrid-Universal
- [x] 7. replace all Filter Conditions with Slickgrid-Universal
- [x] 8. replace all Models with Slickgrid-Universal (enums, interfaces)
- [x] 9. replace all Sort Comparers with Slickgrid-Universal
- [x] 10. replace all Services with Slickgrid-Universal
- [x] 11. replace all Extensions with Slickgrid-Universal
   - [x] keep RowDetail Extension since it has specific Angular code
- [x] 12. replace all SASS Stylings with Slickgrid-Universal
- [x] 13. replace all Export to File Service with Slickgrid-Universal Export to File monorepo package
- [x] 14. replace all Export to Excel Service with Slickgrid-Universal Export to Excel monorepo package
- [x] 15. replace all Export to OData Service with Slickgrid-Universal OData monorepo package
- [x] 16. replace all Export to GraphQL Service with Slickgrid-Universal GraphQL monorepo package
- [x] 17. remove `sg` prefix from all events in the global grid options and make sure all events still work
   - also note that data arguments are now passed as Custom Event Detail and must be pulled via `($event.detail)`
- [x] 18. add a Translate Service wrapper of @ngx-translate for Slickgrid-Universal to work properly
- [x] 19. add an EventPubSub Service wrapper to work properly
- [x] 20. upgrade to latest @ngx-translate version
- [x] 21. add demo to cover Column generics (modified Example 2 to demo it)
- [x] 22. use Bootstrap 5 as a base instead of BS3
- [x] 23. drop internal copy of 3rd party lib `multiple-select.js` and use [multiple-select-modified](https://github.com/ghiscoding/multiple-select-modified) instead
- [x] 24. add `yarn.lock` file into Git and also move from CircleCI to GitHub Actions
- [x] 25. create/update [Migration Guide](/ghiscoding/Angular-Slickgrid/wiki/Migration-to-3.x) Wiki to next version `3.x`
- [x] 26. update all necessary Wikis
   - Export to Excel from `@slickgrid-universal/excel-export`
   - Export to Text File from `@slickgrid-universal/text-export`
   - OData Backend Service from `@slickgrid-universal/odata`
   - GraphQL Backend Service from `@slickgrid-universal/graphql`
   - add new Composite Editor Wiki
- [x] 27. rename/move `master` branch into `version2` branch to keep ref to older version
- [x] 28. create a Git Tag on Angular-Slickgrid-Demos of `2.x` version before rewriting to newer version
- [x] 29. jQueryUI is imported separately in Slickgrid-Universal (via `main.ts`), check to see if we need to do that here
- [x] 30. add the new SlickGrid Composite Editor Example (rename previous Example 30 to 31 to align with Aurelia-Slickgrid)
- [x] 31. search for any leftover of `TODO`, `@ts-ignore` and/or `@deprecated` code
- [x] 32. all Cypress E2E tests should work
- [x] 33. all Jest Unit Tests should work and we should still be at 100% test coverage
- [x] 34. replace Slick Empty Component with Slickgrid-Universal monorepo component package
- [x] 35. ~look at replacing Pagination Component with Slickgrid-Universal monorepo component package~
   - won't work unless I drop the component html file and rewrite all DOM elements with pure JS (via `createElement`) and that would be way too long to do, so I won't proceed with this one
- [x] 36. look at replacing Custom Footer Component with Slickgrid-Universal monorepo component package
- [x] 37. rewrite Resizer Service and make it accessible via Slickgrid-Universal common package and use it in Angular-Slickgrid
- [x] 38. ~remove `dist` publish structure and use regular folder structure~
   - doesn't seem possible because of the lib output structure produced by `ng-packagr` and how it then loads in Angular
- [x] 39. make sure `ng-packgr` lib build still work
- [x] 40. make sure that regular Prod build still work
- [x] 41. update all Angular-Slickgrid demos
- [x] 42. add list of optional Slickgrid-Universal packages on the Readme
- [x] 43. investigate why Flatpickr shows a console warning on the Angular-Slickgrid Demo with a single Locale
- [x] 44. apply fix for new bug found with Tree Data
